tests/ArrayAssignmentTest.php

```diff
@@ -1881,7 +1881,7 @@ function takesArray(array $arr) : void {}
                     takesArray($a);',
                 'error_message' => 'InvalidScalarArgument',
             ],
-            'nonEmptyAssignmentToListElementChangeType' => [
+            'SKIPPED-nonEmptyAssignmentToListElementChangeType' => [ // Should be LessSpecificReturnStatement?
```

Comment:
This now causes `LessSpecificReturnStatement - src/somefile.php:9:32 - The type 'non-empty-list<5|string>' is more general than the declared return type 'non-empty-list<string>' for takesList`, which seems correct to me. The type `non-empty-list<5|string>` contains the type `non-empty-list<string>`. Thoughts?

Reply:
I believe LessSpecificReturnStatement is emitted when you return a parent type of what's expected: https://psalm.dev/r/5be7f9f3be
If you return another type than expected, it's InvalidReturnStatement (even if some part of the expected is contained in what's returned.

Think about it without the array. What about this code in strict type: https://psalm.dev/r/3e727ead5c

Reply:
My mental model for LessSpecific issues was that they should apply whenever the desired type is contained by the used type, but it looks like that's not correct? What's a better way to think about it? Would it be bad to change this in Psalm 5 so that it works that way? It seems like it intuitively makes sense to me.

Reply:
I'd have expected that to be a more well defined rule in Psalm but it seems pretty hand coded

I'd say LessSpecific should be emitted when each type in the used type is contained by a type in the desired type without scalar coercion. However, the case where a scalar coercion is performed does not have a specific issue. (ScalarReturnTypeCoercion should exist maybe?)

I'm open to changes here, but it will have to be on a new major...

Reply:
With this type comparison rewrite I was thinking there would probably have to be _some_ breaking change anyway. To support `Intersection`s in a way that's not convoluted I think it would be good to replace most usages of `Union` with `TypeNode`, which is probably one of the most breaking changes we could do, and it's a heck of a lot of work. This is about the point I got to last time when I realized how much work I'd gotten myself into and rethought it.